### ISSUE TITLE
Allow per-content-type customization of author and date display (a.k.a.: "submitted_by in core")

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#2379]
Currently, like in Drupal, there is just one option: on/off, for displaying the author and date on a content type. Ideally we'd be able to change the wording and the date format. Perhaps allow tokens in the field.

Someone made a Drupal contrib module to change it: https://www.drupal.org/project/submitted_by which might provide useful idea on how to approach it.

![](https://www.drupal.org/files/images/submitted_by-fieldset.png)

---

~~PR by @BWPanda: https://github.com/backdrop/backdrop/pull/2780~~
PR by @BWPanda and @stpaultim https://github.com/backdrop/backdrop/pull/3126

---

Issue Advocate: @stpaultim

**Comment by klonos** [comment]
> Someone made a Drupal contrib module to change it:

...and @BWPanda has already ported this to Backdrop: https://github.com/backdrop-contrib/submitted_by 😉

**Comment by herbdool** [comment]
Thanks @klonos, I hadn't seen that. I wonder if it's a good candidate for core as well.

**Comment by herbdool** [comment]
@klonos I see from that module's issue queue that it's not currently working and there's an open PR. The maintainer doesn't seem responsive.

**Comment by herbdool** [comment]
Now that I've looked at submitted_by more closely, it's a bit of a hack. Would be nice to have it better integrated with core.

**Comment by klonos** [comment]
Yeah, the Drupal version was built with the contrib Token as a requirement. Token is in Backdrop core now, so of course there might be a better way to do things.

**Comment by mikemccaffrey** [comment]
I'd really like to see this be configurable as a pseudo field under the manage display tab.

**Comment by BWPanda** [comment]
And that pseudo field should be configurable with options to choose either a short or long value, which can be configured on the Display Settings vertical tab:

![firefox_screenshot_2017-05-05t05-59-50 421z](https://cloud.githubusercontent.com/assets/2385329/25735089/09412fe2-31ac-11e7-83e4-010ed534ccb6.png)

Not as flexible as having a different value for every view mode like Submitted By does, but I think that's a bit of overkill anyway.

**Comment by Dinornis** [comment]
👍 That would be very useful.

**Comment by herbdool** [comment]
If it's a pseudo field then the display settings could be on the view modes instead of on the main content type settings.

**Comment by BWPanda** [comment]
Yes, I like the @mikemccaffrey's and @herbdool's idea of a pseudo field with configuration there instead!

**Comment by BWPanda** [comment]
I started looking into modifying core to allow for a pseudo field for the author/date info, but realised that pseudo/extra fields cannot have settings...

I did come across this however, http://www.joshwaihi.com/content/working-extra-fields-drupal-7, which seems to indicate that it's possible to get around it.

So what's the best way forward?

1. Add the ability for pseudo fields to specify settings (e.g. edit `hook_field_extra_fields()`)
2. Override the display form to add settings for the new pseudo field (e.g. implement `node_form_field_ui_display_form_alter()` (`hook_form_FORM_ID_alter`))

**Comment by BWPanda** [comment]
Create a [feature request](https://github.com/backdrop/backdrop-issues/issues/3859) to allow pseudo fields to have settings, so if that can get built into core, this can progress as such.

**Comment by BWPanda** [comment]
Since the pseudo field idea is currently blocked by https://github.com/backdrop/backdrop-issues/issues/3859, I've made a PR that simply adds per-content-type customisation of the submitted by info for nodes and comments: https://github.com/backdrop/backdrop/pull/2780

**Comment by olafgrabienski** [comment]
@BWPanda, I've had a look at your PR in the sandbox site, very useful, and it looks quite good. I've

- added the custom format `d. F Y` and called it "Posts",
- changed the *Author/date format* for Posts to `[node:created:posts], by [node:author]`, and
- created a test post, *Author/date info* is "31. July 2019, by admin", as expected.

**Comment by stpaultim** [comment]
I would like to revive this issue. I notice that the sandbox is not currently working. Can someone close and reopen this PR to rebuild the sandbox. Let's also try and get some more eyes on this.

**Comment by BWPanda** [comment]
@stpaultim PR reset.

**Comment by klonos** [comment]
I really like this feature, and the code in the PR generally LGTM 👍, except a minor concern re passing input through `filter_xss_admin()` (someone else should validate this - but I think that we should, since we are allowing user input).

My only other concern at this point is that we are introducing a config string which is not translatable at this point. We do have config translation in core since 1.12, but only for "static" config things - not for "dynamic" config (see #3515) ...it does seem that this is the case for the content type description as well. Oh well 🤷‍♂ ...I guess 😅

...in other words, we should be able to do something like this:

```php
'#default_value' => token_replace(config_get_translated('node.type' . $type->name, 'node_submitted_format')),
```

```php
'#default_value' => $type->settings['node_submitted_format'],
```

**Comment by stpaultim** [comment]
I've added myself as the advocate for this issue and added it to todays agenda. I think that this could be a big win and it appears pretty close to being ready. My understanding is that it simply needs code review and some feedback.

**Comment by stpaultim** [comment]
@klonos - Did you get answers to any of the questions you asked in your last comment, during the meeting last week?

**Comment by bd0bd** [comment]
Well, I do not use Backdrop CMS any more and I am not going to use it in future.
Drupal 7 is better suitable for business.
Bye Backdrop!

**Comment by jenlampton** [comment]
I really want this feature. Many of my sites need to change the submitted by line, and needing to do it in the theme seems like an unnecessary complexity.

> I started looking into modifying core to allow for a pseudo field for the author/date info, but realised that pseudo/extra fields cannot have settings...

We are running into this same problem in a few other places: https://github.com/backdrop/backdrop-issues/issues/4031, and https://github.com/backdrop/backdrop-issues/issues/1441. It seems like psuedo fields need to have settings. I'm not sure if this problem is a blocker for this issue anymore, but it will be resolved soon*. Is that approach preferable? What would the differences be between that and this PR?

I do have some concerns about this approach:

1) Most of my sites need to change all submitted by lines so they are consistent. They appear on all content types, and on all comments. Is there a global "Default" setting in this PR, or does every node type need to be managed separately? 

2) We had discussed cleaning up the submitted line once in the past, but got stuck because the change would break existing themes. I'm not sure this approach would have the same result, but we should test it :)

**Comment by jenlampton** [comment]
Thinking about this more... I'm also concerned about adding the use of tokens (a fairly advanced concept) onto the node edit form (intended for everyone, including people who won't use tokens anywhere else). 

I would feel better about that if we could put the results of the token replacement on the form, in the same way that we put the date format on the date formats page:
<img width="877" alt="Screen Shot 2020-03-19 at 1 47 58 PM" src="https://user-images.githubusercontent.com/397895/77113547-45a51f80-69e8-11ea-850c-155e55ba1976.png">

**Comment by quicksketch** [comment]
> I would feel better about that if we could put the results of the token replacement on the form, in the same way that we put the date format on the date formats page:

This can be tricky to do because we need a node in order to do the token replacements. On the node type form, there's an additional challenge in that the node type itself might not exist yet when adding a new node type.

> Is there a global "Default" setting in this PR, or does every node type need to be managed separately?

Providing options for default vs. overrides usually results in a confusing UI. Other settings that are per-node type (like comment or menu settings) don't provide a default vs override, it's just per content type. I think we should keep with that pattern and not introduce the complexity of two levels of settings.

**Comment by jenlampton** [comment]
> On the node type form, there's an additional challenge in that the node type itself might not exist yet when adding a new node type.

I'm not so worried about the node-type form, as that's more of an admin task and I expect those people to know how to work a token.

> we need a node in order to do the token replacements
Can we build a psuedoNode with the pieces we need?

> I think we should keep with that pattern and not introduce the complexity of two levels of settings.

I think this will be fine. We still have the theme layer to do a global override, and that's what we need to do now anyway.

**Comment by quicksketch** [comment]
Code-wise, this is looking great. There are few suggestions in the PR that need to be added (XSS checks) and there is a failing test. I'll take a look at it and see if I can get it passing.

**Comment by quicksketch** [comment]
On further investigation, it appears that calling `filter_xss_admin()` on the token string should not be necessary, because `token_replace()` already sanitizes tokens by default. So fixing the test is about the only thing here. It is currently failing because previously we were outputing a user's name using `theme('username')`. Now with a token, it is output slightly differently.

Previously for anonymous comments, the name would be output as `Jane Smith (not verified)`. With the use of tokens, it was just printing out as `Jane Smith` instead.

A similar difference has occurred for logged-in users. Previously the author's name was a link to their profile. Now it's only the name itself, no link. Honestly I think that may be an improvement, but it's not an intentional change.

To maintain real consistency with the current behavior, we'd need to make a new token that output the user's name run through `theme('username')`. I'd suggest something like `[node:author:name-formatted]` or `[node:author:username]` (the current token is `[node:author:name]`), but that starts to get pretty lengthy.

I also encountered an issue in that _both_ core themes Basis and Bartik do _not_ use the `$submitted` variable in their comment templates. They both individually print out the author and date:

Bartik:
```html
      <div class="submitted">
        <p class="commenter-name"><?php print $author; ?></p>
        <p class="comment-time"><?php print $created; ?></p>
        <p class="comment-permalink"><?php print $permalink; ?></p>
      </div>
```

Basis:

```html
    <div class="attribution">
      <?php print $user_picture; ?>

      <div class="submitted">
        <p class="commenter-name"><?php print $author; ?></p>
      </div>
    </div> <!-- /.attribution -->
  </header> <!-- /.comment-header -->

  <div class="comment-text">

    <div class="comment-title">
      <?php print render($title_prefix); ?>
      <h3><?php print $title; ?></h3>
      <?php print render($title_suffix); ?>
      <?php if ($new): ?>
        <span class="marker"><?php print $new; ?></span>
      <?php endif; ?>
      <span class="comment-time"><?php print $created; ?></span>
    </div>
```

Which means that the setting for formatting the submitted line for comments doesn't do anything if you're using the core themes.

![image](https://user-images.githubusercontent.com/100206/79524212-c91c5580-8014-11ea-8499-b33819b27288.png)

With this being the case, should we provide this setting at all?

**Comment by quicksketch** [comment]
I pushed a commit to the current PR at https://github.com/backdrop/backdrop/pull/2780 that should fix the failing test.

**Comment by klonos** [comment]
During this week's dev meeting we discussed how to move this forward in a way that would make it possible to have this feature in for 1.16. We've decided that it is best to get this feature in only for the node "submitted by", and leave the part about the same feature for comments for a follow up.

The reason for this decision was that the way core themes are currently implemented does not take advantage of the feature when it comes to comments. The negative side-effect of that would be that people would configure the setting for comments in the admin UI, but they would not see it take effect on their site (unless they used a contrib theme that uses the submitted-by variable). Since we cannot change the markup/output of core themes mid-cycle (it would break existing sites), that would mean postponing till 2.0.

**Comment by stpaultim** [comment]
@BWPanda - Do you have time to work on this issue before feature freeze?

Do you need any more information. We discussed this during the most recent dev call. https://www.youtube.com/watch?v=I9D_a3vmHxI

**Comment by stpaultim** [comment]
@jenlampton or @quicksketch 

Do I just have to remove the changes to:

core/modules/comment/comment.module 
core/modules/comment/comment.theme.inc
core/modules/comment/comment.tokens.inc

I could probably handle that.

**Comment by stpaultim** [comment]
OK. I just pushed a modified PR that simply removes the changes made by @BWPanda to the comment module and am about to test it - additional eyes on this would be appreciated. 

https://github.com/backdrop/backdrop/pull/3126

**Comment by quicksketch** [comment]
Looks good to me. Looks trivial with just one half of it. If anyone wants to give this another review, that would be appreciated.

**Comment by herbdool** [comment]
Thumbs up. Looked at code and tested.

**Comment by quicksketch** [comment]
Thanks @herbdool! I've merged https://github.com/backdrop/backdrop/pull/3126 into 1.x for 1.16.0. Thanks to our leads on this issue @stpaultim and @BWPanda!
